### DISCUSSION TITLE
Any clue about what Burn Color is?

### DISCUSSION THREAD

**Discussion by Trainmaster9977** [ScanMountGoat/Smush-Material-Research#39] [Q&A]
Hello! I am actually nearing the completion of my annotation/simplification of the shaders (checked for accuracy in renderdoc) and while I hope to have it out a bit after I rewrite/release my custom shader, there is one annoying unknown standing in my way.

....burn color comes up a LOT, and I have no idea what it means.

It's used a lot with the rim stuff, but given I have no idea what it means, and I wasn't even able to find it used in wgpu last timed I checked, it's surprisingly hard to name the variables.

Like, obviously, when I know what something is meant for, it's easier to name variables. If I know this value is the input specular stuff before its transformed, I can say InputSpec or something. The variables usually end up more complicated by some point but still.

But well uh....with burn color. I mean I know what its doing math wise, but given the staggering amount of times its used with rim, and me not knowing what it's actually FOR, I'm kinda at a loss for figuring out things like variable names. 

Sorry if this is a bit rambly, I just wanted to get this out ASAP so hopefully you'll see it fairly soon, and maybe give me a hint so I can finish up annotating. Thanks!

**Comment by ScanMountGoat** [comment]
I haven't looked into burn color yet. What are you trying to do that requires a fully annotated shader? I didn't start researching the nushdb until well into the development of ssbh_wgpu, so I wouldn't expect to see many of the in game names in model.wgsl. I'd be happy to review any insights you discover on the shader variables or parameters, but it sounds like this doesn't line up with your use case. You don't need to fully understand the code if you only want to edit a small portion.

If you just want to edit shaders, you either need a compiler or some documentation of the program binary format used by the Tegra X1 GPU. I would imagine that an SDK has a shader compiler, but this would create some obvious legal issues. The other way to compile shaders is through a graphics API call. I believe blujay was able to compile basic GLSL shaders with NVN, but I don't know if he ever open sourced any of his work. Compiling your own small shader from scratch is the better than using assembly in my opinion. The other option is to modify the shader assembly itself. While there may be docs somewhere, I wouldn't expect to find readily available resources like you can find for x86 or ARM64. You could try looking at the decompiler code for Ryujinx to get an idea of the format. Modifying the assembly isn't impossible, but it's definitely not as straightforward as you may think.

**Reply by ScanMountGoat** [reply]
As with anything in reverse engineering or programming in general, start with something trivial. An example would be identifying how texture lookups work in assembly and then changing the lookup to use a different texture (it's probably an integer handle). You could also look into changing constants. If simple edits like that don't work yet with your method, then there's no hope of making fully custom shaders.

**Reply by Trainmaster9977** [reply]
Oh, way ahead of you. I've already gone through a long and arduous trial of figuring out how to edit shaders, and have already done a proof of concept. I plan to do an "official" reveal of my methods a bit after I release my cell shader, but if you are really interested, I've talked about most of it on the server. And in fact, I have done a proof of concept! Below are pictures of a modified shader of mine, that took in_NL, compared it with a PRM input, and then used either the COL or NOR texture based on that. 

I know I've talked to you before, and that I can sometimes seem way too idealistic, asking for things that are unrealistic, but well...

I put in the work. And I did it.

![image](https://user-images.githubusercontent.com/97861450/230693961-d7ec5446-619a-419c-8dad-dafa13a86796.png)
![image](https://user-images.githubusercontent.com/97861450/230693938-e04ebe01-e1e6-490e-a543-125291c938e2.png)
![image](https://user-images.githubusercontent.com/97861450/230693973-4d104e69-0d96-4346-8989-f5b30d6e6355.png)

**Reply by ScanMountGoat** [reply]
Great! Nice to see something working in game.
